### ISSUE TITLE
Guzzle Dependency / Websocket Requirements

### ISSUE THREAD

**Issue by ChadSikorra** [maclof/kubernetes-client#78]
Instead of having Guzzle and Ratchet as hard dependencies, should there instead be other interfaces used so we can use other HTTP clients or websocket clients?

For HTTP clients this is a little more straight forward since [PSR-18](https://www.php-fig.org/psr/psr-18/) exists for that reason. For websockets, something else would have to be made since there is no standardization, but there are competing clients.

One of my use cases for a separate websocket client is so I could use the Swoole websocket client in place of ratchet. Could potentially just create an interface in this library, and then separate out the ratchet implementation into its own repo. Thoughts?

**Comment by maclof** [comment]
@ChadSikorra I completely agree and have been thinking about using the PSR standard instead of a hard dependency on Guzzle

The problem does arise with the websocket client, I do not currently have the time to abstract that implementation out, if there was an existing standard we could use it would be ideal.

**Comment by maclof** [comment]
HTTP transport now uses an injectable PSR-18 client, so Guzzle is no longer hard-wired there. WebSocket upgrades remain on Ratchet because there is still no equivalent PHP standard; we do not plan a separate abstraction at present. Closing with the HTTP portion implemented.
